Makefile

```diff
@@ -1,175 +1,389 @@
 DEFAULT_GOAL: help
 SHELL := /bin/bash
 GCLOUD_VERSION := 222.0.0-1
-PWD := $(shell pwd)
+SDROOT := $(shell pwd)
 TAG ?= $(shell git rev-parse HEAD)
 STABLE_VER := $(shell cat molecule/shared/stable.ver)
+VENV := $(shell readlink -m .venv)
+VENV2 := $(shell readlink -m .venv2)
+
+SDBIN := $(SDROOT)/securedrop/bin
+DEVSHELL := $(SDBIN)/dev-shell
+
+
+######################################
+#
+# Python environments and dependencies
+#
+######################################
+
+.PHONY: venv
+venv: ## Provision a Python 3 virtualenv for development.
+	@echo "███ Preparing Python 3 virtual environment..."
+	@PYTHON_VERSION=3 $(SDROOT)/devops/scripts/boot-strap-venv.sh
+	@echo
+
+.PHONY: venv2
+venv2: ## Provision a virtualenv for compiling Python 2 requirements.
+	@echo "███ Preparing Python 2 virtual environment..."
+	@PYTHON_VERSION=2 $(SDROOT)/devops/scripts/boot-strap-venv.sh
+	@echo
+
+.PHONY: update-admin-pip-requirements
+update-admin-pip-requirements: ## Update admin requirements.
+	@make -s -C admin update-pip-requirements
+
+.PHONY: update-python3-requirements
+update-python3-requirements: venv  ## Update Python 3 requirements with pip-compile.
+	@echo "███ Updating Python 3 requirements files..."
+	@$(VENV)/bin/pip-compile \
+		--output-file securedrop/requirements/python3/develop-requirements.txt \
+		admin/requirements-ansible.in \
+		securedrop/requirements/python3/develop-requirements.in
+	@$(VENV)/bin/pip-compile \
+		--output-file securedrop/requirements/python3/test-requirements.txt \
+		securedrop/requirements/python3/test-requirements.in
+	@$(VENV)/bin/pip-compile --generate-hashes \
+		--output-file securedrop/requirements/python3/securedrop-app-code-requirements.txt \
+		securedrop/requirements/python3/securedrop-app-code-requirements.in
+
+.PHONY: update-python2-requirements
+update-python2-requirements: venv2 ## Update Python 2 requirements with pip-compile.
+	@echo "███ Updating Python 2 requirements files..."
+	@$(VENV2)/bin/pip-compile \
+		--output-file securedrop/requirements/python2/develop-requirements.txt \
+		admin/requirements-ansible.in \
+		securedrop/requirements/python2/develop-requirements.in
+	@$(VENV2)/bin/pip-compile \
+		--output-file securedrop/requirements/python2/test-requirements.txt \
+		securedrop/requirements/python2/test-requirements.in
+	@$(VENV2)/bin/pip-compile --generate-hashes \
+		--output-file securedrop/requirements/python2/securedrop-app-code-requirements.txt \
+		securedrop/requirements/python2/securedrop-app-code-requirements.in
```

Comment:
For the `pip-compile` calls, we'd do better to wrap those in a container with a specific Python version, as described in #4296. Reasons:

* Developers may not use a predictable location for their venvs (`echo $VIRTUAL_ENV` on my machine shows `/home/user/.virtualenvs/sd`, for instance)
* Directly invoking a script _outside_ whatever the "active" virtualenv is will cause problems. By bottling up in a container, we'll get more predictable behavior on the pip updates.

The second point is critical for app-code dependency updates: we want to use specifically the version of Python used by apt servers, _not_ whatever version of Python is used on a developer laptop.

Reply:
For an example of where we use containerized pip-compile elsewhere, see https://github.com/freedomofpress/securedrop.org/blob/b074ade61d58220adfdd0fb183d3da535388c970/Makefile#L43-L51 For Xenial support, we want Python 3.5.2:

```
vagrant@app-staging:~$ python3 --version
Python 3.5.2
```

Reply:
Added. I chose to do this in our existing Ubuntu 16.04 containers, as they should match production.